python/ray/serve/tests/test_http_state.py

```diff
@@ -3,172 +3,517 @@
 
-        # Unblock health check, so status should transition UNHEALTHY -> HEALTHY
-        signal.send.remote()
-        wait_for_condition(check_proxy, status=HTTPProxyStatus.HEALTHY, timeout=2)
+    # Ensure the _consecutive_health_check_failures is correct
+    assert proxy_state._consecutive_health_check_failures == 3
 
-    ray.shutdown()
 
+@patch("ray.serve._private.http_state.DEFAULT_READY_CHECK_TIMEOUT_S", 1)
+def test_http_proxy_state_update_starting_ready_always_timeout():
+    """Test calling update method on HTTPProxyState when the proxy state is STARTING and
+    when the ready call always timed out.
 
-def test_http_proxy_shutdown():
-    ray.init()
+    The proxy state started with STARTING. After update is called, ready calls takes
+    very long time to finish. The state will eventually change to UNHEALTHY after all
+    retries have exhausted.
+    """
 
     @ray.remote(num_cpus=0)
-    class MockHTTPProxyActor:
+    class NewMockHTTPProxyActor:
+        async def ready(self):
+            await asyncio.sleep(100)
+
+        async def check_health(self):
+            pass
+
+    proxy_state = _create_http_proxy_state(proxy_actor_class=NewMockHTTPProxyActor)
+
+    # Ensure the proxy status before update is STARTING.
+    assert proxy_state.status == HTTPProxyStatus.STARTING
+
+    # Continuously trigger update and wait for status to be changed.
+    wait_for_condition(
+        condition_predictor=_check_proxy,
+        state=proxy_state,
+        status=HTTPProxyStatus.UNHEALTHY,
+    )
+
+    # Ensure the proxy status is changed to UNHEALTHY after update.
+    assert proxy_state.status == HTTPProxyStatus.UNHEALTHY
+
+    # Ensure the _consecutive_health_check_failures is correct.
+    assert proxy_state._consecutive_health_check_failures == 3
+
+
+@patch("ray.serve._private.http_state.PROXY_HEALTH_CHECK_PERIOD_S", 0.1)
+def test_http_proxy_state_update_healthy_check_health_succeed():
+    """Test calling update method on HTTPProxyState when the proxy state is HEALTHY and
+    when the check_health call succeeded
+
+    The proxy state started with HEALTHY. After update is called and ready call
+    succeeded, the status will change to HEALTHY. After the next period of check_health
+    call, the status should stay as HEALTHY.
+    """
+    proxy_state = _create_http_proxy_state()
+
+    # Continuously trigger update. The status should change from STARTING to HEALTHY
+    # when ready.
+    wait_for_condition(
+        condition_predictor=_check_proxy,
+        state=proxy_state,
+        status=HTTPProxyStatus.HEALTHY,
+    )
+    first_check_time = proxy_state._last_health_check_time
+
+    # Ensure the proxy status is HEALTHY after updates.
+    assert proxy_state.status == HTTPProxyStatus.HEALTHY
+
+    # Continuously trigger update and status continue to be HEALTHY.
+    wait_for_condition(
+        condition_predictor=_check_proxy,
+        state=proxy_state,
+        status=HTTPProxyStatus.HEALTHY,
+    )
```

Comment:
I think this returns as soon as the condition is met, so it only checks the status once, not continuously. A simple assert or a short for loop might be more straightforward, depending on what you're trying to check

Reply:
Due to those `ready()` and `check_health()` all running asynchronously, `wait_for_condition()` is required for making sure the state transition is met 😅